website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+
+Executing the code below, we demonstrate that our Multi-Lingual Query Engine is able to perform Natural Language to SQL generation and query execution.
+
+```python
+import os
+
+import mlflow
+from database import setup_database
+from definitions import (
+    EXPERIMENT_NAME,
+    MODEL_ALIAS,
+    REGISTERED_MODEL_NAME,
+    REMOTE_SERVER_URI,
+)
+from dotenv import load_dotenv
+from vector_store import setup_vector_store
+
+mlflow.set_tracking_uri(REMOTE_SERVER_URI)
+mlflow.set_experiment(EXPERIMENT_NAME)
+mlflow.langchain.autolog()
+
+
+def main():
+    # Load environment variables from .env file
+    load_dotenv()
+
+    # Access secrets using os.getenv
+    os.environ["OPENAI_API_KEY"] = os.getenv("OPENAI_API_KEY")
+
+    # Setup database and vector store
+    conn = setup_database()
+    cursor = conn.cursor()
+    vector_store = setup_vector_store()
+
+    # Load the model
+    model_uri = f"models:/{REGISTERED_MODEL_NAME}@{MODEL_ALIAS}"
+    model = mlflow.pyfunc.load_model(model_uri)
+    model_input = {"conn": conn, "cursor": cursor, "vector_store": vector_store}
+    app = model.predict(model_input)
+
+    # save image
+    app.get_graph().draw_mermaid_png(
+        output_file_path="sql_agent_with_safety_checks.png"
+    )
+
+    # Example user interaction
+    print("Welcome to the SQL Assistant!")
+    while True:
+        question = input("\nEnter your SQL question (or type 'exit' to quit): ")
+        if question.lower() == "exit":
+            break
+
+        # Initialize the state with all required keys
+        initial_state = {
+            "messages": [("user", question)],
+            "iterations": 0,
+            "error": "",
+            "results": None,
+            "generation": None,
+            "no_records_found": False,
+            "translated_input": "",  # Initialize translated_input
+        }
+
+        solution = app.invoke(initial_state)
+
+        # Check if an error was set during the safety check
+        if solution["error"] == "yes":
+            print("\nAssistant Message:\n")
+            print(solution["messages"][-1][1])  # Display the assistant's message
+            continue  # Skip to the next iteration
+
+        # Extract the generated SQL query from solution["generation"]
+        sql_query = solution["generation"].sql_code
+        print("\nGenerated SQL Query:\n")
+        print(sql_query)
+
+        # Extract and display the query results
+        if solution.get("no_records_found"):
+            print("\nNo records found matching your query.")
+        elif "results" in solution and solution["results"] is not None:
+            print("\nQuery Results:\n")
+            for row in solution["results"]:
+                print(row)
+        else:
+            print("\nNo results returned or query did not execute successfully.")
+
+    print("Goodbye!")
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
Can you show an example invocation of this (not all the code, just one quick example) toward the beginning?

Reply:
I've included an example invocation early on to help illustrate the workflow. The example demonstrates how a user input like 'Quantos clientes temos por país?' (in Portuguese) is processed step by step, from translation to SQL generation and execution.